### ISSUE TITLE
Recurring but not manually reproducible deadlock with, 100% cpu, no logs

### ISSUE THREAD

**Issue by koenpunt** [puma/puma#1123]
On random times one of the threads of Puma peaks to >100% CPU, and no request are processed.

The weird thing is there are no logs of failure whatsover.

I first had in a cluster mode, but the same issue arises in single mode. 

Running `ltrace` on the hanging process shows me a lot of `pthread_mutex_lock` and `pthread_mutex_unlock` calls. 

``` log
$ sudo ltrace -p 29193
pthread_mutex_unlock(0x83324f10, 0, 0, 0xffffffff)                                                                          = 0
pthread_mutex_lock(0x7d37ff30, 0, 0x83324f10, 0xffffffff)                                                                   = 0
pthread_mutex_unlock(0x7d37ff30, 0, 0, 0xffffffff)                                                                          = 0
clock_gettime(1, 0x663e24c0, 0, 0)                                                                                          = 0
time(0)                                                                                                                     = 1477428260
pthread_mutex_lock(0x821936f0, 0x7cc9d9c0, 0x7cc9d9c0, 0x663e4058)                                                          = 0
pthread_mutex_unlock(0x821936f0, 0x7cc9d9c0, 0, 0x663e4058)                                                                 = 0
pthread_mutex_lock(0x821936f0, 0x7cc9d9c0, 0x7cc9d9c0, 0x83324de0)                                                          = 0
pthread_mutex_unlock(0x821936f0, 0, 0x7c9e8ed0, 0x83324de0)                                                                 = 0
time(0)                                                                                                                     = 1477428260
clock_gettime(1, 0x663e24c0, 0, 0x663e4038)                                                                                 = 0
clock_gettime(1, 0x663e2450, 0x83324f10, 0x33d6ee)                                                                          = 0
_setjmp(0x83324ff8, 0x663e2450, 0x19c64150, 0xffffffff)                                                                     = 0
pthread_mutex_lock(0x7d37ff30, 0, 0xb5baca99, 0x83324de0)                                                                   = 0
pthread_mutex_unlock(0x7d37ff30, 0, 0, 0x83324de0)                                                                          = 0
pthread_mutex_lock(0x83324f10, 0, 0x7d37ff30, 0x83324de0)                                                                   = 0
pthread_cond_timedwait(0x83324e98, 0x83324f10, 0x663e2450, 0x83324de0 <unfinished ...>
pthread_mutex_lock(0x7cca4ae0, 1, 0, 0xffffffff)                                                                            = 0
pthread_kill(0x66ded700, 26, 0, 0xffffffff <no return ...>
--- SIGVTALRM (Virtual timer expired) ---
```

More here: https://gist.github.com/koenpunt/34507228b0af5cb8b75692500a46710b

I've now mitigated the problem by spawning 2 separate Puma processes, and monitoring both with monit, and respawning them if the CPU get above a certain threshold. But this is definitely not optimal.

Maybe someone else can shed some light on this?

**Comment by koenpunt** [comment]
Update; the problem even occurs when running 2 separate processes. Could it have to do with server limits? Because I've seen other issues mentioning that. And if that could be the case, how to track that down?

**Comment by schneems** [comment]
Can you give us an example app that reproduces the problem? Are you running the latest version of puma?

**Comment by steakknife** [comment]
(A minimal example as a public docker container might be ideal for repro.)

**Comment by koenpunt** [comment]
Creating a repro is difficult, because I haven't been able to manually reproduce it. But I might have found a pointer on what is causing it. Will get back when I know more.

**Comment by edwardmp** [comment]
I think I have experienced similar behavior. Puma seems to be in a deadlock and does not recover unless the full docker instance with puma in it is restarted. I do not have 100% CPU utilization though.

I noticed that this issue started occurring after we updated from Puma 3.6 to 3.6.2.
For now I've rolled back to 3.6 and the issue disappeared. Logs did not seem to reveal anything out of the ordinary.

**Comment by nateberkopec** [comment]
@edwardmp You may be experiencing #1085.

**Comment by nateberkopec** [comment]
Feel free to comment/reopen if this is still occuring, but we've done a lot of bug squashing in this area in the last year.

**Comment by vk26** [comment]
I have faced with this problem. The reason was not in the puma, but in the rack. The problem manifested itself when was uploaded large file The problem was solved by upgrade rack ( I had version 2.0.1, updated to 2.0.4). After that, the problem no longer arose.
[Link to issue with rack](https://github.com/rack/rack/issues/1075)
